src/test/java/com/puppycrawl/tools/checkstyle/AbstractModuleTestSupport.java

```diff
@@ -360,10 +362,19 @@ protected final void verify(Checker checker,
             message.append("differing violations: ").append(differingViolations);
         }
 
```

Comment:
Can we do away with all the Maps and simply compare ```actualViolations``` to be equal to ```realExpectedViolations``` instead?

```
        assertWithMessage("Violations should be identical.")
                .that(actualViolations)
                .containsExactlyEntriesIn(realExpectedViolations)
```

Reply:
It is actually ```containsExactlyEntriesIn(``` for maps. I tried but it's not working as expected.

Reply:
>  tried but it's not working as expected.

 I will use maps in my example so that we can discuss the output (though most of the time, `verify()` receives a Map with a single entry, so it would look more like above):

```bash

Violations should be identical.
keys with wrong values
for key:
    MyOtherFile.java
expected value:
    11:17: '42' is a magic number.
    13:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
but got value:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
for key:
    MyFile.java
expected value:
    6: First sentence of Javadoc is missing an ending period.
    9: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
but got value:
    6: First sentence of Javadoc is missing an ending period.
    7: First sentence of Javadoc is missing an ending period.
    17: First sentence of Javadoc is missing an ending period.
    27: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
---
expected:
    MyOtherFile.java:
    11:17: '42' is a magic number.
    13:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
    MyFile.java:
    6: First sentence of Javadoc is missing an ending period.
    9: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.
    
    
but was:
    MyOtherFile.java:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
    MyFile.java:
    6: First sentence of Javadoc is missing an ending period.
    7: First sentence of Javadoc is missing an ending period.
    17: First sentence of Javadoc is missing an ending period.
    27: First sentence of Javadoc is missing an ending period.
    36: First sentence of Javadoc is missing an ending period.

```

Please share what is unexpected here; maybe we could improve this idea.

Reply:
I'll look at it again. I found that it was passing even when I flipped the order of violations in an input file.

Reply:
The ordering was creating a fuss. Fixed and done now.

Reply:
Can you show a failure in CI, and share a link here? It would be good to see failed assertion format.

Reply:
Hm. I now notice that `containsExactlyEntriesIn()` does not give the diff that `containsExactlyElementsIn()`. Perhaps it is better to do:

```java

        expectedViolationsMap.forEach((k,v) -> {
            assertWithMessage("Violations should be identical.")
                .that(v)
                .containsExactlyElementsIn(actualViolationsMap.get(k))
                .inOrder();
        });

```

Yielding:

```bash

Violations should be identical.
missing (1):
    12:17: '42' is a magic number.
unexpected (1):
    18:17: '42' is a magic number.
---
expected:
    11:17: '42' is a magic number.
    12:17: '42' is a magic number.
    14:17: '42' is a magic number.
    
but was:
    11:17: '42' is a magic number.
    14:17: '42' is a magic number.
    18:17: '42' is a magic number.

```

Reply:
But what if the key k does not exist in ```actualViolationsMap```? ```actualViolationsMap.get(k)``` will return NULL.

Reply:
This is why I added 
```java

        assertWithMessage("Files in expected and actual differ.")
            .that(actualViolationsMap.keySet())
            .isEqualTo(expectedViolationsMap.keySet());

```
to my example :)

Reply:
Got it

Reply:
done

Reply:
> The entire idea of Truth is to make assertions more fluent and English-like.

I agree with this 100%. I put comments in my example for discussion's sake; you should remove them. I think that our intention is very clear now.

Instead of `key` and `value`, what do you think about `filename` and `violationList`?

Instead of  `assertWithMessage("Violations should be identical.")`, what do you think about including the filename somehow? In the case of multiple files, it might not be easy to determine which file has differing violations. Maybe `Violations for 'MyFile.java' differ:`?

Reply:
@shashwatj07 @nmancus1 Intellij offers to compare the strings and view the differences as shown in [viewing-and-exploring-test-results.html](https://www.jetbrains.com/help/idea/viewing-and-exploring-test-results.html#assert-equals) . It comes really handy when there is a number of huge violation messages. This feature might be in other IDEs as well. Right now the "difference viewer" option is not showing up. Can we integrate it into our code as well?

Something like 

```
// check violations
realExpectedViolations.forEach((filename, violations) -> {
    assertWithMessage("Violations in \"%s\" should be identical.", filename)
        .that(actualViolations.get(filename).toArray())
        .isEqualTo(violations.toArray());
});
```

The output remains same just I added the file name to message:
```
Violations in "H:\GitHub\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\checks\imports\unusedimports\InputUnusedImportsCheckClearState.java" should be identical.
expected        : [3:8: Unused import - java.util.Arrays., 4:12: Unused import - java.util.List., 5:9: Unused import - java.util.Set.]
but was         : [3:8: Unused import - java.util.Arrays., 4:8: Unused import - java.util.List., 5:8: Unused import - java.util.Set.]
differs at index: [1]
Expected :[3:8: Unused import - java.util.Arrays., 4:12: Unused import - java.util.List., 5:9: Unused import - java.util.Set.]
Actual   :[3:8: Unused import - java.util.Arrays., 4:8: Unused import - java.util.List., 5:8: Unused import - java.util.Set.]
```

I am getting the "difference viewer" option with this change, also it should not change the test as well according to my understanding. Please take a look and share your views.

Reply:
@nmancus1 sure, done!

Reply:
@AkMo3 From the link it seems that comparing differences is only available for ```string assertEquals failures```. Is it so?

Reply:
@shashwatj07 No. `assertEquals` is just an example. As you can see below, I am getting _**click to see difference**_ option
![image](https://user-images.githubusercontent.com/69365878/120080266-a98ba380-c0d5-11eb-938b-baef8fda933b.png)

Reply:
> Please take a look and share your views.

@AkMo3 I think the current implementation addresses the comments at https://github.com/checkstyle/checkstyle/issues/9142#issuecomment-848891011 and takes advantage of Truth's awesome assertion output (thus making the 'migration' complete). I do not hold want to hold up this PR any longer :)

@pbludov implementation has changed significantly, please take another look at this PR. What do you think about overriding `toString()` to make output like https://github.com/checkstyle/checkstyle/pull/10026#discussion_r641577310 (multiple lines) ?

Reply:
This can be resolved.